extensions/amp-ad/0.1/amp-ad-custom.js

```diff
@@ -0,0 +1,121 @@
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+import {templatesFor} from '../../../src/template';
+import {xhrFor} from '../../../src/xhr';
+
+/** @const {!string} Tag name for custom ad implementation. */
+export const TAG_AD_CUSTOM = 'amp-ad-custom';
+
+/** @var {Object} A map of promises for each value of data-url. The promise
+ *  will fetch data for the URL for the ad server, and return it as a map of
+ *  objects, keyed by slot; each object contains the variables to be
+ *   substituted into the mustache template. */
+const ampCustomadXhrPromises = {};
+
+/** @var {Object} a map of ad slot ids for each value of data-url.
+ *  Each entry in the map is an array of slot ids. */
+let ampCustomadSlots = null;
+
+export class AmpAdCustom extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+    /** @private {string} The base URL of the ad server for this ad */
+    this.url_ = element.getAttribute('data-url');
+
+    /** @private {string} A string identifying this ad slot: the server's
+     *  responses will be keyed by slot */
+    this.slot_ = element.getAttribute('data-slot');
+    if (!this.slot_) {
+      // If a slot wasn't specified, set it to zero. For use in pages
+      // with only a single ad.
+      this.slot_ = 0;
+    }
+  }
+
+  /** @override */
+  getPriority() {
+    // Loads ads after other content.
+    return 2;
+  }
+
+  /** @override **/
+  isLayoutSupported(layout) {
+    /** @TODO Add proper support for more layouts, and figure out which ones
+     *  we're permitting */
+    return isLayoutSizeDefined(layout);
+  }
+
+  buildCallback() {
+    // Ensure that there are templates in this ad
+    const templates = this.element.querySelectorAll('template');
+    user().assert(templates.length > 0, 'Missing template in custom ad');
+    // And ensure that the slot value is legal
+    user().assert(this.slot_.match(/^[0-9a-z]+$/),
+        'custom ad slot should be alphanumeric: ' + this.slot_);
+  }
+
+  /**
+   * Get a URL which includes a parameter indicating all slots to be fetched
+   * from this web server URL
+   * @private getFullUrl_
+   * @returns {String} The URL with the "ampslots" parameter appended
+   */
+  getFullUrl_() {
+    if (ampCustomadSlots === null) {
+      // The array of ad slots has not yet been built, do so now.
+      ampCustomadSlots = {};
+      const elements = document.querySelectorAll('amp-ad[type=custom]');
+      for (let index = 0; index < elements.length; index++) {
+        const elem = elements[index];
+        const url = elem.getAttribute('data-url');
+        if (!(url in ampCustomadSlots)) {
+          ampCustomadSlots[url] = [];
+        }
+        const slotId = elem.getAttribute('data-slot');
+        ampCustomadSlots[url].push(slotId ? slotId : 0);
+      }
+    }
+    return this.url_ + (this.url_.match(/\?/) ? '&' : '?') + 'ampslots=' +
```

Comment:
We have a [`addParamToUrl`](https://github.com/ampproject/amphtml/blob/master/src/url.js#L149-L152) to do this. Also, we need to `encodeURIComponent` the ad slot's joined string.